eclipse-theia/theia
# Add dynamic debug configurations API

#### What it does


Fixes: #8652

This initial implementation exposes plugin provided dynamic debug configurations via the quick pick command palette.
e.g.
`ctrlcmd+p -> type 'debug ' (Note the space at the end).`

or alternatively: 

`ctrl+shft+p -> type 'Select and start debugging'`

If an extension provider of dynamic debug configurations is loaded, these entries are listed and available for selection.

#### How to test

#Alternative 1
1) Assuming the extension `ms-vscode.js-debug` is installed as a builtin dependency
Open a `javascript` or `typescript` project, build it  and open in editor a `.js` or `.ts` file ready to be debugged.
> ctrlcmd+p -> type 'debug ' (Note the space at the end).
2) You will notice two new entries under the `pwa-node` debug type, 
Select the one called `Run current file`
3) Verify that the debug session starts and that debug options work (e.g. stop at break points, etc.).

![run_current_file](https://user-images.githubusercontent.com/76971376/133849565-ee7bb602-faff-4f06-8d35-06a91d0ce159.gif)

NOTE: The dynamic configuration **'Create JavaScript debug terminal',** does not work properly, this will be tracked in the following issue #9978.

#Alternative 2 (For a more in depth testing and development)

The following URL points to a fork repository of `vscode-mock-debug`, and the `theia-dynamic-debug` branch is provided for test purposes of this functionality.

* from your `theia` development repository install the `vscode-mock-debug` extension inside the `plugins` folder via the following commands:
> git clone https://github.com/alvsan09/vscode-mock-debug.git; cd vscode-mock-debug; git checkout  theia-dynamic-debug
> yarn
> yarn compile
> yarn run vscode:prepublish

* open the following file (from the extension above) in the editor: 
 > plugins/vscode-mock-debug/sampleWorkspace/readme.md

* From the command palette type 'debug ' or use the shortcut keys below:
> ctrlcmd+p -> type 'debug ' (Note the space at the end).

* select one of the provided dynamic configurations under `mock`:
![image](https://user-images.githubusercontent.com/76971376/129972347-f4c10206-34af-46a0-945c-0760da831ff7.png)

* Example:
![dynamic_mock_debug](https://user-images.githubusercontent.com/76971376/129973298-f8a20ff9-cbdf-4921-b2dc-791b3bc809e5.gif)

#### Review checklist


#### Reminder for reviewers

- as a reviewer, I agree to behave in accordance with [the review guidelines](https://github.com/theia-ide/theia/blob/master/doc/pull-requests.md#reviewing)

Signed-off-by: Alvaro Sanchez-Leon <alvaro.sanchez-leon@ericsson.com>
